src/main/webapp/app/shared/modal/AddMutationModal.tsx

```diff
@@ -80,507 +82,747 @@ function AddMutationModalContent({
+    setTabStates(states => {
+      const newStates = _.cloneDeep(states);
+      newStates.splice(tabIndex, 1, ...newAlterations);
+      return newStates;
+    });
+  }
 
-      if (firstIteration) {
-        for (const excluded of excluding) {
-          excludingPromises.push(flowResult(annotateAlteration({ alteration: excluded, geneIds: [geneEntity.id] }, false)));
-        }
-      }
-      firstIteration = false;
-    }
-    const allAlterations = await Promise.all([...alterationPromises, ...excludingPromises]);
-    const alterations = allAlterations.slice(0, alterationPromises.length);
-    const excludingAlterations = allAlterations.slice(alterationPromises.length);
+  const fetchNormalAlterationDebounced = useCallback(
+    _.debounce((newAlteration: string, tabIndex: number, alterationData: AlterationData[]) => {
+      fetchNormalAlteration(newAlteration, tabIndex, alterationData);
+    }, 1000),
+    [tabStates.length]
+  );
 
-    const newAlterations = alterations.map((alteration, index) => {
-      return {
-        label: mutationOptionsToAdd[index].label,
-        value: mutationOptionsToAdd[index].label,
-        uuid: mutationOptionsToAdd[index].uuid,
-        Alteration: mutationOptionsToAdd[index].label,
-        Name: alteration.entity?.name || '',
-        Protein_Change: alteration.entity?.proteinChange,
-        Protein_Start: alteration.entity?.start?.toString() || '',
-        Protein_End: alteration.entity?.end?.toString() || '',
-        Ref_Residues: alteration.entity?.refResidues || '',
-        Var_Residues: alteration.entity?.variantResidues || '',
-        Consequence: alteration.entity?.consequence?.name || '',
-        Comment: comment,
-        Excluding: excludingAlterations.map(excludingAlteration => ({
-          label: excludingAlteration.entity?.alteration,
-          value: excludingAlteration.entity?.alteration,
-          uuid: generateUuid(),
-          Alteration: excludingAlteration.entity?.alteration || '',
-          Name: excludingAlteration.entity?.name || '',
-          Protein_Change: excludingAlteration.entity?.proteinChange,
-          Protein_Start: excludingAlteration.entity?.start?.toString() || '',
-          Protein_End: excludingAlteration.entity?.end?.toString() || '',
-          Ref_Residues: excludingAlteration.entity?.refResidues || '',
-          Var_Residues: excludingAlteration.entity?.variantResidues || '',
-          Consequence: excludingAlteration.entity?.consequence?.name || '',
-          Excluding: [],
-          warning: excludingAlteration.warning ? excludingAlteration.message : null,
-          error: excludingAlteration.error ? excludingAlteration.message : null,
-        })),
-        warning: alteration.warning ? alteration.message : null,
-        error: alteration.error ? alteration.message : null,
-      };
+  function handleNormalAlterationChange(newValue: string, tabIndex: number) {
+    setTabStates(states => {
+      const newStates = _.cloneDeep(states);
```

Comment:
Is it necessary to do deep clone?

Reply:
React recommends treating state as immutable and always making copies: https://react.dev/learn/updating-objects-in-state.

I also just tested, and my component was not rerendering when I mutated the state. I needed to return a new object with a new reference.